agent/src/error.rs

```diff
@@ -0,0 +1,55 @@
+use crate::apiclient::ApiClientError;
+use snafu::Snafu;
+
+/// The crate-wide result type.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// The crate-wide error type.
+#[derive(Debug, Snafu)]
+#[snafu(visibility = "pub")]
+pub enum Error {
+    #[snafu(display("Unable to create client: {}", source))]
+    ClientCreate { source: kube::Error },
+
+    #[snafu(display(
+        "Unable to fetch the Kubernetes custom resource associated with this node: {}",
+        source
+    ))]
+    FetchCustomResource { source: kube::Error },
+
+    #[snafu(display("Unable to get associated node name: {}", source))]
+    GetNodeName { source: std::env::VarError },
+
+    #[snafu(display(
+        "Unable to fetch the Kubernetes node associated with this pod: {}",
+        source
+    ))]
+    FetchNode { source: kube::Error },
+
+    #[snafu(display("Uable to get Node uid because of missing"))]
+    MissNodeUid {},
+
+    #[snafu(display("Uable to get Bottlerocket node 'status' because of missing"))]
+    MissBottlerocketNodeStatus,
+
+    #[snafu(display("Uable to get Bottlerocket node 'spec' because of missing"))]
+    MissBottlerocketNodeSpec,
+
+    #[snafu(display("Unable to gather system version metadata: {}", source))]
+    BottlerocketNodeStatusVersion { source: ApiClientError },
+
+    #[snafu(display("Unable to gather system available versions metadata: {}", source))]
+    BottlerocketNodeStatusAvailableVersions { source: ApiClientError },
+
+    #[snafu(display("Unable to gather system available versions metadata: {}", source))]
+    UpdateBottlerocketNodeResource { source: reqwest::Error },
+
+    #[snafu(display("Unable to take action `prepare`: {}", source))]
+    PrepareUpdate { source: ApiClientError },
+
+    #[snafu(display("Unable to take action `update`: {}", source))]
+    PerformUpdate { source: ApiClientError },
+
+    #[snafu(display("Unable to take action `reboot`: {}", source))]
+    BootUpdate { source: ApiClientError },
```

Comment:
Maybe this could be the same error, and just receive `what` as a parameter:

```suggestion
    #[snafu(display("Unable to take action '{}': {}", action, source))]
    # Maybe a better name, this is just a suggestion
    TakeAction { action: String, source: ApiClientError },
```